File: algorithms/powerOfTwo/PowerOfTwo.cpp
// Source : https://leetcode.com/problems/power-of-two/
// Author : Hao Chen
// Date   : 2015-07-16

/********************************************************************************** 
 * 
 * Given an integer, write a function to determine if it is a power of two.
 * 
 * Credits:Special thanks to @jianchao.li.fighter for adding this problem and creating 
 * all test cases.
 *               
 *               
 *               
 **********************************************************************************/


class Solution {
public:

    // count the number fo bits 1, if it only has one, then return true
    bool isPowerOfTwo01(int n) {
        int cnt = 0; //num of bits 1
        for(; n>0; n>>=1){
            if ( n & 1 ) {
                cnt++;
                if (cnt>1) return false;
            }
        }
        return cnt==1;
    }
    
    //we notice: 2^n - 1 always be 1111111...
    //so, (2^n) & (2^n-1) always be zero
    bool isPowerOfTwo02(int n) {
        return n<=0 ? false : (n & (n-1)) == 0;
    }
    
    bool isPowerOfTwo(int n) {
        if (random()%2){
            return isPowerOfTwo02(n);
        }
        return isPowerOfTwo01(n);
    }
};

File: algorithms/java/src/PowerOfTwo/PowerOfTwo.java
// Source : https://leetcode.com/problems/power-of-two/
// Author : Diego Ruiz Piqueras (Pikeras72)
// Date   : 24-04-2022

/***************************************************************************************************** 
 *
 * Given an integer n, return true if it is a power of two. Otherwise, return false.
 * 
 * An integer n is a power of two, if there exists an integer x such that n == 2x.
 *
 * Example 1:
 * 
 * Input: n = 1
 * Output: true
 * Explanation: 2^0 = 1
 * 
 * Example 2:
 *
 * Input: n = 16
 * Output: true
 * Explanation: 2^4 = 16
 * 
 * Example 3:
 * 
 * Input: n = 3
 * Output: false
 * 
 * Constraints:
 * 
 *  1 <= n <= 2^31 - 1
 * 
 * Explanation of the solution:
 * 
 * After doing the log2(n), if the result have no decimal part, it means that 'n' is a power of two.
 ******************************************************************************************************/

class Solution {
    public boolean isPowerOfTwo(int n) {
        double num =Math.log10(n)/ Math.log10(2);
        if(num % 1 != 0){
            return false;
        }
        return true;
    }
}
